Add model export utilities for YOLOv8 and YOLOv9

spal-synaptics opened the pull request:
- Added export script to export YOLO model to tflite / onnx / torch / tensorflow + generate YAML metafile for onnx and tflite
- Added convert script to convert exported model to SyNAP
  - Currently using the SyNAP docker, will upgrade to the pip library once #3 is merged
- Added scripts to copy models to board and profile with `synap_cli`
  - Currently via SSH, will add ADB support in the future

See the quickstart in /model/README.md for instructions on using these scripts.

spal-synaptics committed:
add code for model export

spal-synaptics committed:
update to exclude model export files

spal-synaptics committed:
add documentation for model export

spal-synaptics committed:
update YAML processor from PyYAML to ruamel.yaml

spal-synaptics committed:
changed to cpu-only version for PyTorch

spal-synaptics committed:
add instance segmentation masks overlay (#4)

spal-synaptics committed:
fix program path in argparse prog argument

spal-synaptics committed:
update cleanup functions to be more modular

spal-synaptics committed:
add export files cleanup function

spal-synaptics committed:
add caching for model export files

spal-synaptics committed:
exclude model export cache

Meet-synaptics commented on model/README.md:
```diff
@@ -0,0 +1,90 @@
+| [YOLOv9](https://docs.ultralytics.com/models/yolov9/) | [export/yolo.py](/model/export/yolo.py) |
+
+## YOLO Export Quickstart
+This tutorial will go through exporting the YOLOv8 segmentation (small) model into two input sizes (640x353, 224x224) and two quantization types (uint8, float16). The quantization dataset used is [`coco8-seg`](https://docs.ultralytics.com/datasets/segment/coco8-seg/) and is assumed to be located at `"/home/$USER/synaptics-synap/toolkit/datasets/coco8-seg/*.jpg"`. The board used is the Astra SL168 ...
```
typo 640*353 => 640*352

Meet-synaptics commented on model/README.md:
```diff
@@ -0,0 +1,90 @@
+
+| Model | Export Script |
+|-------|---------------|
+| [YOLOv8](https://docs.ultralytics.com/models/yolov8/) | [export/yolo.py](/model/export/yolo.py) |
```
What are the limitations here to generalize this tool for any version of YOLO model?

spal-synaptics replied:
The current limitation is that the model export has to happen through the ultralytics package, specifically through `Model.export()`. So we can't do, for example, YOLOv5 as that's it's own repository and not through the ultralytics package.
That being said, the model exporter is modular. So someone could just subclass the base `ModelExporter` and provide their own `export_model()`  for any model. All they need to ensure is that `export_model()` should return the path to the exported tflite/onnx/torch etc. model.

Meet-synaptics commented on model/README.md:
````diff
@@ -0,0 +1,90 @@
+### 1. Export YOLOv8 to TFLite models
+```
+python -m model.export.yolo \
+    --models yolov8s-seg \
````
can it export classification/object-detection/pose models?

spal-synaptics replied:
Yep, should be able to export all models supported by ultralytics's `Model.export()`.

Meet-synaptics commented:
We could also rename 'model' to something 'export_scripts' which might be a more descriptive name

Meet-synaptics commented on model/requirements.txt:
```diff
@@ -0,0 +1,81 @@
+absl-py==2.1.0
```
We also need to exclude this all requirements.txt files while building the pip package. The current implementation of the wheel build includes all requirements.txt files within the toolkit repository. I think we can avoid including these requirements in the Python wheel?

spal-synaptics committed:
fix typo in model input size

Meet-synaptics committed:
add pip configuration (#3)

* Toolkit 3.1.1 - Added pip support

* modify workflow

* Simplified requirement.txt
- Updated wheel package name to conform with PEP440 naming standards.
- Update install_require for wheel package

* Simplified requiremnet.txt
- Update wheel package name as per PEP440 Naming standards
- Update install_require for wheel package
- Update github action to upload wheels in specified draft release

* Update workflow

- Remove deprecation warning for Node.js
- Build only binary distribution

* Added 'image_od', 'image_from_raw', and 'image_to_raw' tools to the wheel
- Excluded unnecessary files from the wheel
- Incremented toolkit version
- Added console script for the 'help' command

* Modify help console script

spal-synaptics committed:
add cache cleanup

spal-synaptics committed:
update requirements to exclude redundant dependencies

spal-synaptics committed:
Merge branch 'main' of github.com:synaptics-synap/toolkit into wip/model-export

spal-synaptics committed:
update error message

spal-synaptics committed:
update to synap pip package

spal-synaptics committed:
add ADB support

spal-synaptics committed:
update docs for ADB support in copy.py and profile.py

spal-synaptics committed:
add common functions for getting input and output info

spal-synaptics committed:
add more output info (name, shape) in metadata

spal-synaptics committed:
update to more consistent formatting

spal-synaptics committed:
add partial metadata generation for not fully supported model types

mdenna-synaptics reviewed (approved):
looks good to me, the only comment as already suggested by Meet is to rename "model" folder to something more descriptive of the content, such as "export".

spal-synaptics commented:
> looks good to me, the only comment as already suggested by Meet is to rename "model" folder to something more descriptive of the content, such as "export".

@mdenna-synaptics @Meet-synaptics How about "model_tools" as a name? I can't use "export" as that's already a sub-module and the module does do non-export tasks like copying models to the board and profiling them.

mdenna-synaptics commented:
> > looks good to me, the only comment as already suggested by Meet is to rename "model" folder to something more descriptive of the content, such as "export".
> 
> @mdenna-synaptics @Meet-synaptics How about "model_tools" as a name? I can't use "export" as that's already a sub-module and the module does do non-export tasks like copying models to the board and profiling them.

I see.. or maybe just "tools" or "utils" would also look good to me..
